runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryTimerInternals.java

```diff
@@ -155,10 +156,18 @@ public void setTimer(TimerData timerData) {
   @Override
   public void deleteTimer(
       StateNamespace namespace, String timerId, String timerFamilyId, TimeDomain timeDomain) {
-    throw new UnsupportedOperationException("Canceling a timer by ID is not yet supported.");
+    TimerData removedTimer = existingTimers.remove(namespace, timerId + '+' + timerFamilyId);
+    if (removedTimer != null) {
+      Preconditions.checkState(
+          removedTimer.getDomain().equals(timeDomain),
+          "%s doesn't match time domain %s of timer",
+          timeDomain,
+          removedTimer.getDomain());
+      timersForDomain(timeDomain).remove(removedTimer);
+    }
   }
 
-  /** @deprecated use {@link #deleteTimer(StateNamespace, String, TimeDomain)}. */
+  /** @deprecated use {@link #deleteTimer(StateNamespace, String, String, TimeDomain)}. */
```

Comment:
Maybe this should defer to the `deleteTimer(StateNamespace, String, String, TimeDomain)` implementation?

Reply:
I considered doing that, though it would mean an additional lookup the get the `TimeDomain` first. That didn't seem worth it ... But no real preference from my side.